Add support for permission contexts

pop4959 opened the pull request:
As discussed in #2960. Allows Essentials to register contexts with permission plugins that support them. A working implementation for LuckPerms is included in this PR. PermissionsEx 2 also has support for registering contexts, however it should probably be added at a later date, since as per @zml2008 on Discord, the API is not yet stable.

![image](https://user-images.githubusercontent.com/17698576/82718550-f5507500-9c57-11ea-970d-9d5f8bce8037.png)

pop4959 committed:
Add permission contexts to Essentials

zml2008 commented:
Thanks for getting this going pop -- it'll be nice to have EssentialsX doing contexts.

However, i'm not super enthusiastic about the way you've abstracted contexts. At the level in this PR, any new permissions plugin would have to reimplement every context EssX provides.

Instead, what do you think about having a higher-level abstraction, so that individual contexts are on the EssX side of the abstraction, not an implementation detail. The API I'm envisioning would look something like:

```java

interface IPermissionsHandler {
    // [...]
    void registerContext(String key, BiConsumer<User, Consumer<String>> calculator, Set<String> suggestableValues);
    // [...]
}
```

Then, registering a context (using the AFK context as an example) would look like:

```java

handler.registerContext("afk", (user, result) -> result.accept(String.valueOf(user.isAfk())), ImmutableSet.of("true", "false"));
```

What do you think? That change feels like it would result in an API that's a lot more suited to future growth, while not exposing too many LP/PEX implementation details to the plugin.

pop4959 committed:
More abstraction

pop4959 commented:
@zml2008 The only reason why I didn't was because I was trying to figure out how permission plugins are doing contexts, and wasn't trying to make any assumptions. As far as with LuckPerms & PEX 2, what you say makes a lot of sense because they both utilize consumers to find out what the context is. I don't know if there are any other plugins that do contexts, but as long as they all work in a similar way, it does make a whole lot of sense to abstract it more on Essentials' side. Check out the last commit I pushed and let me know if that looks a bit better.

zml2008 commented:
Hm, I think that's a lot better. 

This doesn't affect any of what the context abstraction is being used for at the moment, but again thinking down the road, having the calculation callback be a `BiConsumer` allows for EssX contexts where multiple values may be applicable to a player at once -- and that is a fairly fundamental part of the "specification" for contexts (as much as one doesn't really exist... oh well)

pop4959 commented:
Yeah it's a little hard to tell since there isn't a hard specification on it. I do like how it works in both LuckPerms and PEX though, so hopefully other plugins follow the existing conventions set between the two. Ideally contexts could be supported through a common API such as Vault, however unless that's something that is known to come in the foreseeable future, I think it's worth handling it in Essentials. I mean, there was a time where Essentials had a different permissions implementation for a dozen or so plugins... at least here most everything comes from Vault, and contexts are the only special case (for now).

lucko commented:
Super cool! Awesome to see other plugins integrating contexts! 😎 

I agree with zml:

> but again thinking down the road, having the calculation callback be a BiConsumer allows for EssX contexts where multiple values may be applicable to a player at once 

As well as allowing for multiple, doing this ^ would also allow registering a calculator which (occasionally) supplies no contexts at all for a player.

pop4959 commented:
Ok, so I understand what you're going for, but I'm having some trouble envisioning how this would actually be implemented. I changed `BiConsumer<User, Consumer<String>>` from the suggested implementation to `Function<Player, String>` since the `calculate` method of `ContextCalculator` is responsible for figuring out what contexts are applicable, and it expects values to be fed into its `ContextConsumer`. Right now values (a single value) can be fed into it by applying the function to the player it is attempting to get contexts for. Maybe I'm missing something or thinking about this the wrong way, but I don't know how BiConsumer is useful in this situation, since it doesn't produce any values. Any advice would be appreciated.

lucko commented:
Hopefully this helps to demonstrate:

```java
public void setup() {
    registerContext(
            "vanished",
            (player, consumer) -> consumer.accept(Boolean.toString(ess.getUser(player).isHidden())),
            ImmutableSet.of("true", "false")
    );
}

public void registerContext(String key, BiConsumer<Player, Consumer<String>> valueFunction, Set<String> suggestions) {
    luckPerms.getContextManager().registerCalculator(new ContextCalculator<Player>() {
        @Override
        public void calculate(Player target, ContextConsumer consumer) {
            valueFunction.accept(target, value -> consumer.accept(key, value));
        }
    });
}
```

or an approach using Collections would work too.

```java
public void setup() {
    registerContext(
            "vanished",
            player -> Collections.singleton(Boolean.toString(ess.getUser(player).isHidden())),
            ImmutableSet.of("true", "false")
    );
}

public void registerContext(String key, Function<Player, Iterable<String>> valueFunction, Set<String> suggestions) {
    luckPerms.getContextManager().registerCalculator(new ContextCalculator<Player>() {
        @Override
        public void calculate(Player target, ContextConsumer consumer) {
            valueFunction.apply(target).forEach(value -> consumer.accept(key, value));
        }
    });
}
```

(omitted `estimatePotentialContexts` for the sake of brevity)

Turbotailz commented:
Super cool! Are there plans to add more? There are a few that could probably be done, like flying and god mode but the most important one I can think of is when a player is jailed.

zml2008 commented:
oh thanks turbo -- that also brings up: if we use jails as a context following the original issue where the context value is the jail name, the `Set` of suggested values should probably turn into a `Supplier<Set>`, so values can change if jails are added/removed at runtime

Turbotailz commented:
Have two contexts in that case, maybe? `jailed` boolean and `jail` name.

pop4959 committed:
Apply suggestions

pop4959 commented:
> Hopefully this helps to demonstrate

@lucko Thanks for the example! I definitely see where I got confused. I've thought about it a bit now, and opted for the second approach because I think the code is more readable (less trying to figure out which consumer is doing what).

> Super cool! Are there plans to add more? There are a few that could probably be done, like flying and god mode but the most important one I can think of is when a player is jailed.

@Turbotailz Yes, almost certainly. For clarification, the focus of this PR is adding *support* for contexts, not exhaustively adding every possible context that could be desired. I've gone ahead to implement a few simple ones that I thought were useful. More can be added later, and I'm sure there will be feature requests for contexts that no one thought of.

> the `Set` of suggested values should probably turn into a `Supplier<Set>`, so values can change if jails are added/removed at runtime

@zml2008 Good point, I glanced over this as well. Changes applied!

> Have two contexts in that case, maybe? `jailed` boolean and `jail` name.

@Turbotailz Yep, I was thinking this as well. For jails, people may wish to just check if a player is jailed at all, or if they are in a specific jail. I'm not sure how popular the jails feature of Essentials is though, so I didn't include those contexts in this PR. Although, I'm sure *someone* will ask for them, so they'll probably get tacked on at some point.

pop4959 committed:
iterable instead of set

BrainStone commented:
One common request for a context we get a lot on the LP discord is gamemode.

zml2008 commented:
hm, that seems like a useful context, but imo data that's part of the vanilla game is within the scope of permissions plugins -- such as world, world type, gamemode, and EssX should only provide contexts to expose its own functionality

BrainStone commented:
Considering it overrides the commands and takes a certain amount of control over it, I think it's justified.  
And Luck's take for LuckPerms has always been that that sort of thing should be implemented through external plugins. While I do not personally agree with that notion, I do understand his reasoning. I also do think that essentials is a good candidate to implement it too. There should be a check to make sure the context doesn't exist already (ExtraContexts adds it too iirc).

pop4959 commented:
I actually agree with both sides here, as it's a bit of a grey area. Features such as afk and muting are clearly Essentials' own functionality, while others like vanish and fly are functionality that many people access through Essentials, but are actually features of Vanilla or Bukkit. So it's not completely out of the question to add such contexts, but we'd want to be mindful of whether it strays too far from the core feature set.

> There should be a check to make sure the context doesn't exist already (ExtraContexts adds it too iirc).

This is a question I had while implementing this, which I have yet to figure out. If two plugins register contexts with the same name, what happens? It could be a problem, especially since many people use Essentials for the wide variety of general commands it has, but then use a specialized plugin to override certain features. For example, a moderation plugin may request to register a "muted" context, and in such a circumstance, Essentials would not want to register its own. I don't see how this is handled in the LuckPerms API, as `ContextCalculator` doesn't appear to be aware of what context it is calculating.

mbax commented:
> vanish and fly are functionality that many people access through Essentials, but are actually features of Vanilla or Bukkit

Nitpick: Vanish is a plugin thing. Per-player visibility is a Bukkit thing. :D

lucko commented:
If you think collisions are going to be an issue, then just namespace them I guess.

e.g. for WorldGuard regions I currently use `wg-region`, you could use `ess-muted`, etc if you wanted to - but personally I don't think it's worthwhile.

mbax commented:
> If you think collisions are going to be an issue, then just namespace them I guess.

Shouldn't namespaces be a feature of permissions plugins? Like, EssX should just be registering "vanish" and then on the LuckPerms end it's "Essentials:vanish" right? Otherwise you're going to run into some fun collisions down the line.

lucko commented:
> Shouldn't namespaces be a feature of permissions plugins?

Yes, probably should be. I wasn't thinking that far ahead when designing the system though unfortunately.

Ichbinjoe reviewed (dismissed):
Talked in Discord, these contexts should really be at least fake namespaced (prefixed with 'ess-' or something idk pop figure it out)

mbax commented:
Namespace with plugin name to encourage predictability!

pop4959 committed:
Address concerns about reloads

pop4959 committed:
Support context prefixing

mbax commented on Essentials/src/com/earth2me/essentials/Settings.java:
```diff
@@ -1653,6 +1653,12 @@ public boolean logCommandBlockCommands() {
+    // #easteregg
+    @Override
+    public String getContextPrefix() {
+        return config.getString("context-prefix", "");
```
Just want to note my objection here to defaulting to no namespacing! :)

pop4959 replied:
Noted! I went and explained my reasoning below. The issue of default namespacing can be tackled at a later date, assuming that it becomes a real issue.

pop4959 commented:
Ok, so due to the concerns on Discord about reloading the plugin, I went ahead and fixed things up so that contexts are properly unregistered on disable. I tested reloading prior to these changes and found no issues with contexts continuing to work, however it's probably best to be on the safe side of things.

Also added a context prefix setting. For now, this is hidden from users by default, as there are no known conflicts with any other plugins' contexts, nor should it generally be an issue. Server admins can change the prefix if they need to. If it becomes a major issue in the future, I think it would make sense to change the default to a suitable prefix. But for now, I don't think it's necessary, and I think having nice easy to read contexts is much friendlier to users.

kashike commented:
> Also added a context prefix setting. For now, this is hidden from users by default, as there are no known conflicts with any other plugins' contexts, nor should it generally be an issue. Server admins can change the prefix if they need to. If it becomes a major issue in the future, I think it would make sense to change the default to a suitable prefix. But for now, I don't think it's necessary, and I think having nice easy to read contexts is much friendlier to users.

Personally, I think the prefix should be static. `ess-` is a short and simple prefix, being only one character longer than `wg-`, and allowing users to configure their own does not make much sense to me - regular players would never see this, and no conflicts would exist.

pop4959 committed:
Revert "Support context prefixing"

This reverts commit 85b626cc2ef70c19afceae801cc6d8d6243a969a.

pop4959 committed:
Static prefix

pop4959 commented:
Fair enough, I guess my idea was a bit silly. I got a few more opinions and it looks like the popular choice is to use `ess-` as the prefix, so I've gone ahead and done that for now.

pop4959 committed:
this is probably also good to do

lucko commented on Essentials/src/com/earth2me/essentials/EssentialsPluginListener.java:
```diff
@@ -25,6 +27,9 @@ public void onPluginEnable(final PluginEnableEvent event) {
         }
         ess.getPermissionsHandler().setUseSuperperms(ess.getSettings().useBukkitPermissions());
         ess.getPermissionsHandler().checkPermissions();
+        ess.getPermissionsHandler().registerContext("ess-afk", player -> Collections.singleton(String.valueOf(ess.getUser(player).isAfk())), () -> ImmutableSet.of("true", "false"));
```
These register calls should probably be in `onEnable`??

Unless I'm missing something, they're currently being added on every `PluginEnableEvent`? (this is gonna fire every time a plugin loads?!)

pop4959 replied:
No... I think you're right about that. It is definitely going to fire every time a plugin loads. I actually did originally have these in `onEnable`, but the problem is that before `checkPermissions` is run (which is only done here) the `handler` in `PermissionsHandler` is null. It's probably done in this fashion to avoid adding softdepends on every Vault-based permission plugin.

Two possible solutions:
1) From the event, check if the plugin is a known one that supports contexts.
2) Schedule a task to register contexts on the first server tick.

I don't particularly like either solution, but it doesn't seem like Bukkit has any event that gets called after all plugins are loaded. Maybe something else works too, but this is what comes to mind. Any opinions?

pop4959 replied:
So I got another idea from Glare (on Discord) that I like slightly better, which is a spin-off of my first solution. Instead of checking from this event, we do actually do it in `onEnable`, since after we've checked that a context supporting permission plugin is loaded, it should be safe to register the contexts since Essentials has a softdepend on it, and thus `checkPermissions` has to have run already and have a non-null handler.

lucko replied:
You're also already storing a
```java
private Set<ContextCalculator<Player>> contextCalculators
```

Wouldn't be the biggest change to make that a `Map<String, ContextCalculator<Player>>` and just check to see if already registered - just an idea.

However, I'd suggest re-thinking whether that logic really needs to be running every time a plugin enables in the first place...

lucko replied:
```java
ess.getPermissionsHandler().setUseSuperperms(ess.getSettings().useBukkitPermissions());
ess.getPermissionsHandler().checkPermissions();
ess.getPermissionsHandler().registerContext(...);
```

can all be moved to `onEnable`, you're softdepending on LuckPerms now anyway.

Then just keep a single call to
```java
ess.getPermissionsHandler().checkPermissions();
```
in the plugin enable listener to catch other permission plugins.

The call to
```java
ess.getPermissionsHandler().unregisterContexts();
```

definitely needs to be moved to `onDisable`. In it's current position, contexts are going to be unregistered every time *any* plugin is disabled.

mdcfe replied:
These should only be re-applied when the currently active permissions handler changes.

I'm not sure how I feel about registering these here - I'd personally have put them in a separate method or even a new class, as the amount of contexts will likely increase and it feels cluttered to throw them in here. Perhaps they could be added to a default `registerContexts` method on `PermissionHandler`?

pop4959 replied:
Yeah, it wasn't my intention to leave the registrations there either. It was just an easy place to put them for now while testing everything. They can be registered from anywhere though, whether that's primarily in `onEnable`, or a `registerContexts` method. For the latter, we probably want to consider if all contexts will always be enabled all at once, or if there are contexts which may need to be initialized elsewhere (because the calculator function needs to be defined in scope of whatever it's using, although most of the time that would probably be the `Essentials` instance).

pop4959 committed:
Handle registrations in onEnable/onDisable

pop4959 committed:
Change prefixes again

pop4959 commented:
Changes applied! Just to be clear, I am aware of the clutter with having all of the contexts registered directly in onEnable. Just waiting on a response from @md678685 now about what we want to do there. I don't really think we need yet another method in `PermissionHandler` specifically for this, since (as I mentioned before) it may be the case that not all registrations can easily be made in one place. A private helper method in the Essentials class may do fine, in my opinion. It's only conveniently so that all three contexts so far only require the plugin instance in their calculation.

How it looks now
![image](https://user-images.githubusercontent.com/17698576/82884844-2bf8ea80-9ef9-11ea-849a-c78af782c82b.png)

lucko reviewed (dismissed):
LGTM :)

mdcfe commented:
> > it may be the case that not all registrations can easily be made in one place

@pop4959 When would this actually be the case? I can't think of any EssentialsX features where this is a reasonable concern (at least, under the current architecture of the plugin).

pop4959 commented:
> @pop4959 When would this actually be the case? I can't think of any EssentialsX features where this is a reasonable concern (at least, under the current architecture of the plugin).

¯\\\_(ツ)_/¯

I think it mostly just comes down to preference at this point, which is why I asked what you'd like to do about it. We could put it in a `registerContexts` method in `PermissionHandler`, as you suggest. It would need the plugin instance as a parameter, and it can register all/most contexts. The main thing I just wanted to point out was that the design of the interface for contexts in Essentials allows them to be registered easily from anywhere in the plugin. I do not know when this would be the case, but I think it is useful to consider.

mdcfe commented:
> 
> 
> > @pop4959 When would this actually be the case? I can't think of any EssentialsX features where this is a reasonable concern (at least, under the current architecture of the plugin).
> 
> ¯\_(ツ)_/¯
> 
> I think it mostly just comes down to preference at this point, which is why I asked what you'd like to do about it. We could put it in a `registerContexts` method in `PermissionHandler`, as you suggest. It would need the plugin instance as a parameter, and it can register all/most contexts. The main thing I just wanted to point out was that the design of the interface for contexts in Essentials allows them to be registered easily from anywhere in the plugin. I do not know when this would be the case, but I think it is useful to consider.

If there was anywhere that these could be registered organised by feature then perhaps this would be useful, but as it stands EssentialsX doesn't currently do anything else feature-by-feature. I'd personally register them all in a new method on the `PermissionHandler` class (not `IPermissionHandler`) since that's where the rest of the permissions system setup is organised. They could also be registered in `onEnable` but I'm not a fan since there isn't any other permissions setup there.

One concern is that right now, EssentialsX can change the permissions backend at runtime, and in this PR's current state contexts will only be registered to the permissions handler that was detected when the plugin first enables. This means if a permissions plugin loads after EssentialsX for some reason, the contexts will never be applied (I believe this is why they were originally in `PluginEnableEvent` which is where we listen for new permissions plugins). I'm not sure what the best course of action is here - we could listen specifically for LP/PEX/other context plugins to enable in case they end up enabling late, but this feels awkward (would mean more ties to the permissions plugins in places we don't necessarily want them).

pop4959 commented:
> If there was anywhere that these could be registered organised by feature then perhaps this would be useful, but as it stands EssentialsX doesn't currently do anything else feature-by-feature. I'd personally register them all in a new method on the `PermissionHandler` class (not `IPermissionHandler`) since that's where the rest of the permissions system setup is organised. They could also be registered in `onEnable` ...

Sure, I mean we could have a `public PermissionHandler::initContexts(Essentials)` or something, which is just called in `onEnable`, instead of just dumping it in there directly. This could also just as easily be a helper method such as `private Essentials::initContexts()` Just a matter of how to organize it.

> ... but I'm not a fan since there isn't any other permissions setup there.

I think this needs some clarification. Contexts are not the same as permissions. They are used for managing permissions, but they do need to be set up by Essentials. Permissions themselves are set up and provided by permission plugins. (also btw hat permissions are actually registered in `onEnable` apparently)

> One concern is that right now, EssentialsX can change the permissions backend at runtime, and in this PR's current state contexts will only be registered to the permissions handler that was detected when the plugin first enables. This means if a permissions plugin loads after EssentialsX for some reason, the contexts will never be applied (I believe this is why they were originally in `PluginEnableEvent` which is where we listen for new permissions plugins). I'm not sure what the best course of action is here - we could listen specifically for LP/PEX/other context plugins to enable in case they end up enabling late, but this feels awkward (would mean more ties to the permissions plugins in places we don't necessarily want them).

This goes back to the original design decision of adding a `LuckPermsHandler` class, which contains the implementation of Essentials contexts specific to LuckPerms. Currently, the only known permission plugins known to bother with implementing a context system is LuckPerms and PermissionsEx 2. Two implementations isn't too difficult to manage, especially when most of the functionality is provided by Vault (the handlers still extend `ModernVaultHandler`). If and when contexts are popular enough that more than just the two major permission plugins support them, it would make a lot of sense for common API to be included in plugins such as Vault so that the implementation per plugin can be managed more efficiently. Both plugins so far implement contexts in much the same way, so it can be abstracted pretty reasonably. In fact, I played around with implementing PermissionsEx contexts prior to submitting this PR. Nothing much should need to be changed to support it.

mdcfe commented:
@pop4959

> Contexts are not the same as permissions. They are used for managing permissions, but they do need to be set up by Essentials. Permissions themselves are set up and provided by permission plugins.

As far as previous discussions go, contexts are assumed to be a feature of permissions plugins only, and in this PR, context support is done entirely through EssentialsX's permissions abstraction. EssentialsX most definitely *does* set up its own permission abstraction layer.

> (also btw hat permissions are actually registered in onEnable apparently)

Those are Bukkit permissions defaults - they exist outside our permissions layer.

> This goes back to the original design decision [...]

Indeed, this PR does abstract over the permissions implementations. The concern I raised is that the handler that's currently in use can change at runtime, and right now this PR does not account for this - it only registers context calculators once when the plugin starts up, so if the active permissions handler changes (which isn't unlikely, given we have a reload command that can change it, and plugins may enable after EssentialsX does) then contexts won't be registered for the new permissions handler. I have no preference as to where this happens (as long as it's sane), but if the handler changes, the contexts must be registered again on the new handler.

pop4959 commented:
@md678685 

> Indeed, this PR does abstract over the permissions implementations. The concern I raised is that the handler that's currently in use can change at runtime, and right now this PR does not account for this [...]

Ah OK, thanks for clarifying. I wasn't sure what you meant before. I'm aware that the permission handler *may* change at runtime, although I thought this was primarily a side effect of Essentials not softdepending on specific Vault implementations... not necessarily a feature. The current implementation makes an assumption that there is only going to be one permission plugin loaded, and that it won't be hot swapped / disabled. If we're going out of our way to support reloading contexts when the permission plugin is changed, then it may be best to check specifically for plugins that support contexts in `onPluginEnable` and `onPluginDisable` instead of only registering them in `onEnable` and `onDisable`. Let me know what you think, because I'd like to get this right.

mdcfe commented:
The issue with this PR right now is that it introduces state (the registration of contexts) into an interface that was previously assumed to have no state (`IPermissionsHandlers` were previously little more than delegates to their respective backends). `checkPermissions` just throws away and replaces permissions handlers whenever it's called, which is fine when we don't need to persist any state, but it's problematic when we introduce the need to remember whether contexts have been registered yet or not.

> it may be best to check specifically for plugins that support contexts in onPluginEnable and onPluginDisable instead of only registering them in onEnable and onDisable.

This would help, but it would split up the permissions provider detection into different places.

The easiest way would probably be to just unregister all contexts at the top of `checkPermissions` then reregister them after detecting the handler. This lets us keep permissions setup organised in the same place while avoiding reinventing the whole permissions handler system.

pop4959 committed:
Change context registrations only when the active permission handler changes
